### PR TITLE
[Issue-168] Define dedicated SMTPException for SMTP communication failures

### PR DESCRIPTION
**Pull request by gaol**
Motivation:

Fixes: #168 

Conformance:

Your commits should be signed and you should have signed the Eclipse Contributor Agreement as explained in https://github.com/eclipse/vert.x/blob/master/CONTRIBUTING.md
Please also make sure you adhere to the code style guidelines: https://github.com/vert-x3/wiki/wiki/Vert.x-code-style-guidelines

This PR tries to introduce a `SMTPException`, representing the failure when communicating with SMTP server.

### CODE REVIEW

**Comment by gaol** [comment]
@vietj would you please review? thanks

File: src/main/java/io/vertx/ext/mail/SMTPException.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  /**
+   * Checks if it is a transient failure, that the reply status code <code>&gt;=</code> 400 and <code>&lt;</code> 500
+   *
+   * @return true if it is a permanent failure, false otherwise.
+   */
+  public boolean isStatusTemporary() {
+    return replyCode >= 400 && replyCode < 500;
+  }
```
Comment (famod) [inline src/main/java/io/vertx/ext/mail/SMTPException.java:88]: Is this terminology of "permanent"/"fatal" and "transient"/"temporary" mandated by the spec/RFC?

I find it a bit confusing that something like "Domain not found" (450) is considered non-permanent. But I guess the logic behind this is that a domain could just be "offline"due to DNS problems or similar?

Secondarily, I personally would find it clearer if those synonyms could be avoided (e.g. "permanent"/"fatal"), better stick with a single term for each case?

**Review by famod — COMMENTED** [review]
I only checked the `SMTPException` for now.

This is a step in the right direction. As a user, I'd still have to check the codes myself but now I don't have to parse it from the exception message anymore! 👍

File: src/main/java/io/vertx/ext/mail/SMTPException.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * This represents an Exception during communication with SMTP server.
+ *
+ * @author <a href="mailto: aoingl@gmail.com">Lin Gao</a>
+ */
+public class SMTPException extends Exception {
```
Comment (vietj) [inline src/main/java/io/vertx/ext/mail/SMTPException.java:27]: could we extend NoStackTraceThrowable ?

File: src/main/java/io/vertx/ext/mail/SMTPException.java (hunk 1 above)
Comment (gaol) [reply src/main/java/io/vertx/ext/mail/SMTPException.java:88]: Yes, they are defined at https://datatracker.ietf.org/doc/html/rfc5321#section-4.2.1:
> 4yz  Transient Negative Completion reply
> 5yz  Permanent Negative Completion reply

so the reply codes that `>= 400 and < 499` are considered as transient failures, and the reply codes that `>=500` are considered as permanent failures.

> I find it a bit confusing that something like "Domain not found" (450) is considered non-permanent. But I guess the logic behind this is that a domain could just be "offline"due to DNS problems or similar?

I guess so :)

> Secondarily, I personally would find it clearer if those synonyms could be avoided (e.g. "permanent"/"fatal"), better stick with a single term for each case?

Yes, I will update the methods name to `isPermanent()` and `isTransient()`.

File: src/main/java/io/vertx/ext/mail/SMTPException.java (hunk 2 above)
Comment (gaol) [reply src/main/java/io/vertx/ext/mail/SMTPException.java:27]: Sure, will do

**Comment by gaol** [comment]
updated according to feedback above, thanks

**Comment by gaol** [comment]
@vietj I am going to merge this unless you have anything else to comment :)

**Pull request merged**